client/client-multi/client-c/src/platform-impl/kaa_tcp_channel.c

```diff
@@ -0,0 +1,1349 @@
+{
+    KAA_RETURN_IF_NIL(context, KAA_ERR_BADPARAM);
+    kaa_tcp_channel_t *channel = (kaa_tcp_channel_t *) context;
+
+    KAA_LOG_INFO(channel->logger, KAA_ERR_NONE, "Kaa TCP channel(%d) hostname resolve failed", channel->access_point.id);
+
+    channel->access_point.state = AP_NOT_SET;
+
+    return KAA_ERR_NONE;
+}
+
+/*
+ * Read uint32 value from buffer.
+ */
+uint32_t get_uint32_t(const char * buffer)
+{
+    return KAA_NTOHL(*(uint32_t *) buffer);
+}
+
+/*
+ * Connect access point.
+ */
+kaa_error_t kaa_tcp_channel_connect_access_point(kaa_tcp_channel_t *self)
+{
+    KAA_RETURN_IF_NIL(self, KAA_ERR_BADPARAM);
+    if (self->access_point.state != AP_RESOLVED)
+        return KAA_ERR_BAD_STATE;
+
+    KAA_LOG_INFO(self->logger, KAA_ERR_NONE, "Kaa TCP channel(%d) connecting to the server...",
+                            self->access_point.id);
+    kaa_error_t error_code = ext_tcp_utils_open_tcp_socket(
+                &self->access_point.socket_descriptor,
+                (kaa_sockaddr_t *) &self->access_point.sockaddr,
+                self->access_point.sockaddr_length);
+
+    KAA_RETURN_IF_ERR(error_code);
+    self->access_point.state = AP_CONNECTING;
+    return KAA_ERR_NONE;
+}
+
+/*
+ * Release access point.
+ */
+kaa_error_t kaa_tcp_channel_release_access_point(kaa_tcp_channel_t *self)
+{
+    KAA_RETURN_IF_NIL(self, KAA_ERR_BADPARAM);
+    kaa_error_t error_code = KAA_ERR_NONE;
+
+    KAA_LOG_INFO(self->logger, KAA_ERR_NONE, "Kaa TCP channel(%d) releasing channel's resources", self->access_point.id);
+
+    if (self->access_point.state == AP_CONNECTED
+            || self->access_point.state == AP_CONNECTING) {
+        ext_tcp_utils_tcp_socket_close(self->access_point.socket_descriptor);
+    }
+
+    self->access_point.socket_descriptor = KAA_TCP_SOCKET_NOT_SET;
+    self->access_point.state = AP_NOT_SET;
+    self->access_point.id = 0;
+
+    if (self->access_point.hostname) {
+        KAA_FREE(self->access_point.hostname);
+        self->access_point.hostname = NULL;
+        self->access_point.hostname_length = 0;
+    }
+    if (self->access_point.public_key) {
+        KAA_FREE(self->access_point.public_key);
+        self->access_point.public_key = NULL;
+        self->access_point.public_key_length = 0;
+    }
+
+    return error_code;
+}
+
+/*
+ * Write to socket sync services.
+ */
+kaa_error_t kaa_tcp_channel_write_pending_services(kaa_tcp_channel_t *self, kaa_service_t *service, size_t services_count)
+{
+    KAA_RETURN_IF_NIL(self, KAA_ERR_BADPARAM);
+
+    KAA_LOG_INFO(self->logger, KAA_ERR_NONE, "Kaa TCP channel(%d) going to serialize %d pending services",
+                        self->access_point.id,
+                        self->pending_request_service_count);
+
+    KAA_RETURN_IF_NIL(services_count, KAA_ERR_NONE);
+
+    char *buffer = NULL;
+    size_t buffer_size = 0;
+    kaa_error_t error_code = kaa_buffer_allocate_space(self->out_buffer, &buffer, &buffer_size);
+    KAA_RETURN_IF_ERR(error_code);
```

Comment:
With current implementation of the kaa_buffer_allocate_space function, this check will always pass.